src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int length;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<T>(null, value, null);
+            tail = head;
+            tail.setPrev(head);
+            head.setNext(head);
+            length++;
+            return;
+        }
+        Node<T> newTail = new Node<T>(tail, value, null);
+        tail.setNext(newTail);
+        newTail.setPrev(tail);
+        tail = newTail;
+        length++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == length || (isEmpty() && index > -1)) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<T>(null, value, head);
+            head.setPrev(newNode);
+            head = newNode;
+        }
+        Node<T> after = getNode(index, true);
+        Node<T> before = after.getPrev();
+        Node<T> newNode = new Node<T>(before, value, after);
+        if (after != null) {
+            after.setPrev(newNode);
+        }
+        if (before != null) {
+            before.setNext(newNode);
+        }
+        length++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index, false).getItem();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> elementToChange = getNode(index, false);
+        T oldItem = elementToChange.getItem();
+        elementToChange.setItem(value);
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> elementToRemove = getNode(index, false);
+        if (index == length - 1) {
+            elementToRemove.getPrev().setNext(null);
+            tail = elementToRemove.getPrev();
+        } else if (index == 0) {
+            elementToRemove.getNext().setPrev(null);
+            head = elementToRemove.getNext();
+        } else {
+            elementToRemove.getNext().setPrev(elementToRemove.getPrev());
+            elementToRemove.getPrev().setNext(elementToRemove.getNext());
+        }
+        length--;
+        return elementToRemove.getItem();
```

Comment:
I'd like to highlight this common mistake advice [here](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#create-a-separate-method-unlinknode-node)